File: csharp/perfect-numbers/PerfectNumbers.cs
using System;
using System.Collections.Generic;
using System.Linq;

public enum Classification
{
    Perfect,
    Abundant,
    Deficient
}

public static class PerfectNumbers
{
    public static Classification Classify(int number)
    {
        if (number < 1)
        {
            throw new ArgumentOutOfRangeException(nameof(number));
        }

        var  aliquotSum = 0;
        for (var i = 1; i <= number / 2; i++)
        {
            if (number % i == 0)
            {
               aliquotSum += i;
            }
        }

        return aliquotSum == number
            ? Classification.Perfect
            : aliquotSum > number
                ? Classification.Abundant
                : Classification.Deficient;
    }
}

File: go/perfect-numbers/perfect_numbers.go
package perfect

import (
	"errors"
)

// Classification Nicomachus' (60 - 120 CE) classification
type Classification int

const (
	// ClassificationPerfect Perfect number
	ClassificationPerfect Classification = iota
	// ClassificationAbundant abundant number
	ClassificationAbundant
	// ClassificationDeficient deficient number
	ClassificationDeficient
)

// ErrOnlyPositive error if number is not positive
var ErrOnlyPositive error = errors.New("input not a positive number")

// Classify classifies a number per Nicomachus' (60 - 120 CE) classification
func Classify(number int64) (Classification, error) {
	if number < 1 {
		return ClassificationDeficient, ErrOnlyPositive
	}

	var aliquotSum int64 = 0
	var i int64
	for i = 1; i <= number/2; i++ {
		if number%i == 0 {
			aliquotSum += i
		}
	}

	if aliquotSum > number {
		return ClassificationAbundant, nil
	} else if aliquotSum < number {
		return ClassificationDeficient, nil
	}
	return ClassificationPerfect, nil
}
